File: Leetcode 2701-2800/2786/2786.py
# Leetcode 2786: Visit Array Positions to Maximize Score
# https://leetcode.com/problems/visit-array-positions-to-maximize-score/
# Solved on 24th of November, 2025
class Solution:
    def maxScore(self, nums: list[int], x: int) -> int:
        """
        Calculates the maximum score achievable by visiting array positions.

        Args:
            nums: A list of integers representing the values at each position.
            x: An integer representing the penalty for switching parity.

        Returns:
            The maximum score achievable.
        """
        maxEven = nums[0]
        maxOdd = nums[0]

        if nums[0] % 2 == 0:
            maxOdd -= x
        else:
            maxEven -= x

        for i in range(1, len(nums)):
            if nums[i] % 2 == 0:
                maxEven = max(maxEven, maxOdd - x) + nums[i]
            else:
                maxOdd = max(maxOdd, maxEven - x) + nums[i]

        return max(maxEven, maxOdd)

File: Leetcode 2701-2800/2786/2786.java
// Leetcode 2786: Visit Array Positions to Maximize Score
// https://leetcode.com/problems/visit-array-positions-to-maximize-score/
// Solved on 24th of November, 2025
class Solution {
    /**
     * Calculates the maximum score achievable by visiting array positions.
     * @param nums The input array of integers.
     * @param x The penalty for switching between even and odd numbers.
     * @return The maximum score.
     */
    public long maxScore(int[] nums, int x) {
        long evenMax = nums[0];
        long oddMax = nums[0];

        if (nums[0] % 2 == 0) {
            oddMax -= x;
        } else {
            evenMax -= x;
        }

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] % 2 == 0) {
                evenMax = nums[i] + Math.max(evenMax, oddMax - x);
            } else {
                oddMax = nums[i] + Math.max(oddMax, evenMax - x);
            }
        }

        return Math.max(evenMax, oddMax);
    }
}